### ISSUE TITLE
JIT crash/assert while compiling Grpc.core codebase

### ISSUE THREAD

**Issue by davidwrighton** [dotnet/runtime#57535]
Repro instructions:
Download https://microsoft-my.sharepoint.com/:u:/p/davidwr/ETDI552L1RtCugwcniHGv5QBVsNi3PDCpXt6y-tWayhvow?e=f4wH4F and unzip it to a directory.

Using that directory as the working directory, run crossgen2 passing the rsp file in the zip as a response file. For example:

`c:\git\runtime\dotnet c:\git\runtime\artifacts\bin\coreclr\windows.x64.Debug\crossgen2\crossgen2.dll @crossgen2repro.rsp`

If run against a Debug crossgen2, it will produce an assert 
```
C:\git\runtime\src\coreclr\jit\loopcloning.cpp:183
Assertion failed 'genTypeSize(genActualType(op1Tree->TypeGet())) == genTypeSize(genActualType(op2Tree->TypeGet()))' in 'Grpc.Core.Internal.SliceBufferSafeHandle:ToByteArray():System.Byte[]:this' during 'Clone loops' (IL size 215)
```
If run against a release crossgen2 it will produce an AV
```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at Internal.JitInterface.CorInfoImpl.JitCompileMethod(IntPtr ByRef, IntPtr, IntPtr, IntPtr, Internal.JitInterface.CORINFO_METHOD_INFO ByRef, UInt32, IntPtr ByRef, UInt32 ByRef)
--------------------------------
   at Internal.JitInterface.CorInfoImpl.CompileMethodInternal(ILCompiler.DependencyAnalysis.IMethodNode, Internal.IL.MethodIL)
   at Internal.JitInterface.CorInfoImpl.CompileMethod(ILCompiler.DependencyAnalysis.ReadyToRun.MethodWithGCInfo, ILCompiler.Logger)
   at ILCompiler.ReadyToRunCodegenCompilation.<ComputeDependencyNodeDependencies>b__36_0(ILCompiler.DependencyAnalysisFramework.DependencyNodeCore`1<ILCompiler.DependencyAnalysis.NodeFactory>)
   at System.Threading.Tasks.Parallel+<>c__DisplayClass33_0`2[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<ForEachWorker>b__0(Int32)
   at System.Threading.Tasks.Parallel+<>c__DisplayClass19_0`1[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<ForWorker>b__1(System.Threading.Tasks.RangeWorker ByRef, Int32, Boolean ByRef)
   at System.Threading.Tasks.TaskReplicator+Replica.Execute()
   at System.Threading.Tasks.TaskReplicator+Replica+<>c.<.ctor>b__4_0(System.Object)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task+<>c.<.cctor>b__271_0(System.Object)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(System.Threading.Thread, System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
   at System.Threading.Tasks.Task.ExecuteEntryUnsafe(System.Threading.Thread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
```

**Comment by AndyAyersMS** [comment]
You can also repro this by grabbing Grpc.Core and Grpc.Core.Api 2.38 from nuget, extracting the DLLs, and running PMI:
```
corerun.exe c:\repos\jitutils\bin\pmi.dll prepall-quiet C:\bugs\r57535\Grpc.Core.dll

DllNotFoundException Grpc.Core.Internal.CommonPlatformDetection::uname (Unable to load DLL 'libc' or one of its dependencies: The specified module could not be found. (0x8007007E))

Assert failure(PID 32692 [0x00007fb4], Thread: 33432 [0x8298]): Assertion failed 'genTypeSize(genActualType(op1Tree->TypeGet())) == genTypeSize(genActualType(op2Tree->TypeGet()))' in 'Grpc.Core.Internal.SliceBufferSafeHandle:ToByteArray():System.Byte[]:this' during 'Clone loops' (IL size 215)

    File: C:\repos\runtime0\src\coreclr\jit\loopcloning.cpp Line: 183
    Image: c:\repos\runtime0\artifacts\tests\coreclr\Windows.x64.Checked\Tests\Core_Root\corerun.exe
```

**Comment by AndyAyersMS** [comment]
Issue is that `LC_Condition::ToGenTree` is trying to set a `TYP_LONG` variable to a `TYP_INT` zero.

**Comment by AndyAyersMS** [comment]
Suspect cloning gets confused because `V01` is a long, but morph optimizes a cast to int-sized:
```
fgMorphTree BB34, STMT00022 (before)
               [000110] ------------              *  JTRUE     void  
               [000109] ------------              \--*  LT        int   
               [000106] ------------                 +--*  LCL_VAR   int    V10 loc9         
               [000108] ------------                 \--*  CAST      int <- long
               [000107] ------------                    \--*  LCL_VAR   long   V01 loc0         

fgMorphTree BB34, STMT00022 (after)
               [000110] -----+------              *  JTRUE     void  
               [000109] J----+-N----              \--*  LT        int   
               [000106] -----+------                 +--*  LCL_VAR   int    V10 loc9         
               [000107] C----+------                 \--*  LCL_VAR   int    V01 loc0         
```
and so likely cloning thinks V01 is an int?

**Comment by AndyAyersMS** [comment]
Not clear yet how to fix this...

**Comment by EgorBo** [comment]
I assume we can't just add the cast back in `ToGenTree` since we're not sure it was zeroextend/signextend

**Comment by EgorBo** [comment]
I wonder if a naive fix would help https://github.com/EgorBo/runtime-1/tree/fix-loop-clone-assert

**Comment by AndyAyersMS** [comment]
Simple repro:
```c#
using System;
using System.Runtime.CompilerServices;

class X
{
    static long z;

    public static int Main()
    {
        z = 10;
        int[] a = F();
        long zz = z;
        int result = 0;
        for (int i = 0; i < (int) zz; i++)        result += a[i];
        return result;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int[] F()
    {
       int[] result = new int[100];
       result[3] = 100;
       return result;
    }
}
```
w/default behavior:
```
 Assertion failed 'genTypeSize(genActualType(op1Tree->TypeGet())) == genTypeSize(genActualType(op2Tree->TypeGet()))' in 'X:Main():int' during 'Clone loops' (IL size 43)
```

**Comment by AndyAyersMS** [comment]
FYI @JulieLeeMSFT -- we need to fix this for .NET 6 but it's not going to make the RC1 deadline. Should have a fix ready by tomorrow.

**Comment by SingleAccretion** [comment]
> The other approach would be to abandon cloning in a case like this, that seems a bit safer perhaps. Testing this now.

Or give up on labeling loops `LPFLG_VAR_LIMIT` (same the for iterator flag I guess) when the "limit" has `GTF_VAR_CAST` on it. Such a "local" is effectively `CAST(int <- LCL_VAR(long))`, it seems unlikely the current machinery would handle it correctly.

(Seems to not produce SPMI diffs too which is nice)

**Comment by AndyAyersMS** [comment]
Bailing out in cloning
```diff
@ -968,7 +968,14 @@ bool Compiler::optDeriveLoopCloningConditions(unsigned loopNum, LoopCloneContext
         else if (loop->lpFlags & LPFLG_VAR_INIT)
         {
             // initVar >= 0
-            LC_Condition geZero(GT_GE, LC_Expr(LC_Ident(loop->lpVarInit, LC_Ident::Var)),
+            const unsigned initLcl = loop->lpVarInit;
+            if (lvaGetDesc(initLcl)->lvType != TYP_INT)
+            {
+                JITDUMP("> Init var V%02u not TYP_INT\n", initLcl);
+                return false;
+            }
+
+            LC_Condition geZero(GT_GE, LC_Expr(LC_Ident(initLcl, LC_Ident::Var)),
                                 LC_Expr(LC_Ident(0, LC_Ident::Const)));
             context->EnsureConditions(loopNum)->Push(geZero);
         }
@@ -992,9 +999,15 @@ bool Compiler::optDeriveLoopCloningConditions(unsigned loopNum, LoopCloneContext
         }
         else if (loop->lpFlags & LPFLG_VAR_LIMIT)
         {
-            unsigned limitLcl = loop->lpVarLimit();
-            ident             = LC_Ident(limitLcl, LC_Ident::Var);
+            const unsigned limitLcl = loop->lpVarLimit();
+
+            if (lvaGetDesc(limitLcl)->lvType != TYP_INT)
+            {
+                JITDUMP("> Limit var V%02u not TYP_INT\n", limitLcl);
+                return false;
+            }

+            ident = LC_Ident(limitLcl, LC_Ident::Var);
             LC_Condition geZero(GT_GE, LC_Expr(ident), LC_Expr(LC_Ident(0, LC_Ident::Const)));

             context->EnsureConditions(loopNum)->Push(geZero);
```
 produces 2 diffs
```
         -57 (-12.26% of base) : 114034.dasm - ILCompiler.Win32Resources.ResourceData:DoResourceDirectoryRead(System.Reflection.Metadata.BlobReader,int,System.Action`3[System.Object, System.UInt32, System.Boolean])
         -97 (-20.86% of base) : 75579.dasm - Microsoft.Cci.MetadataWriter:SerializeParameterInformation(Microsoft.Cci.IParameterTypeInformation,Microsoft.Cci.BlobBuilder):this
```

**Comment by AndyAyersMS** [comment]
In the 75579 case the limit var is a `ushort` and the current jit produces this mixed type tree, which we evidently can tolerate:
```
Adding loop L00 cloning conditions to BB14
    V03 GE 0 & V03 LE V11.Length
Loop cloning condition tree before morphing:
               [000278] ---X--------              *  JTRUE     void  
               [000277] ---X--------              \--*  EQ        int   
               [000275] ---X--------                 +--*  AND       int   
               [000270] ------------                 |  +--*  GE        int   
               [000268] ------------                 |  |  +--*  LCL_VAR   ushort V03 loc0         
               [000269] ------------                 |  |  \--*  CNS_INT   int    0
               [000274] ---X--------                 |  \--*  LE        int   
               [000271] ------------                 |     +--*  LCL_VAR   ushort V03 loc0         
               [000273] ---X--------                 |     \--*  ARR_LENGTH int   
               [000272] ------------                 |        \--*  LCL_VAR   ref    V11 tmp4         
               [000276] ------------                 \--*  CNS_INT   int    0
```

**Comment by AndyAyersMS** [comment]
Similar for 114034, a `ushort`:
```
Adding loop L01 cloning conditions to BB16
    V12 GE 0 & V12 LE V13.Length
Loop cloning condition tree before morphing:
               [000294] ---X--------              *  JTRUE     void  
               [000293] ---X--------              \--*  EQ        int   
               [000291] ---X--------                 +--*  AND       int   
               [000286] ------------                 |  +--*  GE        int   
               [000284] ------------                 |  |  +--*  LCL_VAR   ushort V12 loc9         
               [000285] ------------                 |  |  \--*  CNS_INT   int    0
               [000290] ---X--------                 |  \--*  LE        int   
               [000287] ------------                 |     +--*  LCL_VAR   ushort V12 loc9         
               [000289] ---X--------                 |     \--*  ARR_LENGTH int   
               [000288] ------------                 |        \--*  LCL_VAR   ref    V13 loc10        
               [000292] ------------                 \--*  CNS_INT   int    0
```

**Comment by JulieLeeMSFT** [comment]
> FYI @JulieLeeMSFT -- we need to fix this for .NET 6 but it's not going to make the RC1 deadline. Should have a fix ready by tomorrow.

We can backport to RC1 as servicing.
CC @jeffschwMSFT

**Comment by jeffschwMSFT** [comment]
@JulieLeeMSFT go ahead and port a fix to release/6.0 and we can then consider for RC1
